retrospectives/2019-11-20-worker-manager-overprovisioning.md

```diff
@@ -0,0 +1,79 @@
+* When releasing any new piece of software there will always be unexpected bugs that come to light. Unfortunately, we hit an expensive bug.
+
+The worker-manager issues that we saw had not been previously observed and were not reproducible in testing. There were some previous concerns about its architecture, but nothing that implied it shouldn’t be used in production.
+
+The separation of worker scanning and provisioning meant that issues scanning for workers could result in overprovisioning. In production, where thousands of instances run concurrently, and more are requested all the time across many regions and availability zones, we ran into issues with worker scanning that led worker-manager to dramatically overprovision.
+
+## Timeline
+  - 2019-11-20 18:06 UTC ::::: Tasks are submitted as part of a try push
+  - 2019-11-20 19:09 UTC ::::: The worker-pool is configured to use spot instances, we see this error: "Error calling AWS API: There is no Spot capacity available that matches your request."
+  - 2019-11-20, 19:38 UTC ::::: More pending tasks as part of the try push => more requestedCapacity
+  - 2019-11-20, 19:48 UTC ::::: A change is landed to allow the worker-pool to use on-demand instances
+  - 2019-11-20 19:39 UTC ::::: Another “Error calling AWS API: There is no Spot capacity available that matches your request.” (the last time we see this, indicating that the change to on-demand has happened)
+  - 2019-11-20 19:40 UTC ::::: We spawn 12 instances (each instance has capacity 4)
+  - 2019-11-20 20:01:15 UTC ::::: We’re at runningCapacity 80
+  - 2019-11-20 20:02:32 UTC ::::: We’re back down to 0 runningCapacity
+  - 2019-11-20 20:05 UTC ::::: "Error calling AWS API: We currently do not have sufficient c5.metal capacity in the Availability Zone you requested (us-west-2a). Our system will be working on provisioning additional capacity. You can currently get c5.metal capacity by not specifying an Availability Zone in your request or choosing us-west-2b, us-west-2c, us-west-2d." This error continues for quite some time, repeatedly
+
+## Fallout
+
+* Some large amount of $$$
+* Loss of trust from other teams and management
+* Lots of scramble and stress on our part in the “relaxing” period after our cloudops migration
+
+## Action Items
+
+* [DONE] The worker-pool was disabled and all workers were terminated as soon as the issue was observed.
+* [DONE] Additional logging and error reporting has been added to worker-manager to call attention to overprovisioning.
+  You can use https://gist.github.com/imbstack/b18a697fb3046a90080499a93b0ffe15 to check for overprovisioned worker-pools
+* [DONE] The configuration of worker-runner on generated AMIs has been corrected to use aws-provider.
+* [DONE] Worker-manager should track instances that it has spawned that have not registered and terminate them, as this is indicative of a configuration issue.
```

Comment:
Since workers can also crash after registering, before claiming a task (e.g. due to a new bug) we should really only not terminate workers if they have called [`queue.claimWork`](https://community-tc.services.mozilla.com/docs/reference/platform/queue/api#claimWork) or [`queue.reclaimTask`](https://community-tc.services.mozilla.com/docs/reference/platform/queue/api#reclaimTask) in e.g. the last hour. Any worker that hasn't called one of these endpoints for an hour is costing money, performing no work, and is unavailable for new tasks.

Separately, it is arguable that a worker that breaks the task reclaim contract resulting in a task being resolved by the queue as `exception/claim-expired` should also be terminated by worker manager. Although this isn't as important as the first option, since within an hour it would be caught by the first mitigation.